packages/xrpc-stream/src/server.ts

```diff
@@ -0,0 +1,56 @@
+import { IncomingMessage } from 'http'
+import { WebSocketServer, ServerOptions, WebSocket } from 'ws'
+import { ErrorFrame, Frame } from './frames'
+import logger from './logger'
+
+export class XrpcStreamServer {
+  wss: WebSocketServer
+  constructor(opts: ServerOptions & { handler: Handler }) {
+    const { handler, ...serverOpts } = opts
+    this.wss = new WebSocketServer(serverOpts)
+    this.wss.on('connection', async (socket, req) => {
+      socket.on('error', (err) => logger.error(err, 'websocket error'))
+      try {
+        for await (const frame of handler(req, socket, this)) {
+          if (frame instanceof ErrorFrame) {
+            await new Promise((res, rej) => {
+              socket.send(frame.toBytes(), { binary: true }, (err) => {
+                if (err) return rej(err)
+                res(undefined)
+              })
+            })
+            throw new DisconnectError()
+          } else {
+            socket.send(frame.toBytes(), { binary: true })
+          }
+        }
+      } catch (err) {
+        if (err instanceof DisconnectError) {
+          return socket.close(err.code ?? CloseCode.Policy)
```

Comment:
We might consider passing along the error code from the error frame to the second arg of `socket.close()`.  Would likely involve `DisconnectError` taking a second constructor argument for it.